tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,310 @@
+    sync::{
+        atomic::{fence, AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+use tracing_core::{
+    field::{FieldSet, Visit},
+    span::{self, Id},
+    Event, Field, Interest, Metadata, Subscriber,
+};
+
+#[derive(Debug)]
+pub struct Registry {
+    spans: Arc<Slab<Data>>,
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+#[derive(Debug)]
+pub struct Data {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    children: Vec<Id>,
+    ref_count: AtomicUsize,
+    // TODO(david): get rid of these
+    values: Mutex<HashMap<&'static str, Value>>,
+}
+
+impl Data {
+    fn drop_ref(&self, ordering: Ordering) -> bool {
+        let refs = self.ref_count.fetch_sub(1, ordering);
+        assert!(
+            if std::thread::panicking() {
+                // don't cause a double panic, even if the ref-count is wrong...
+                true
+            } else {
+                refs != std::usize::MAX
+            },
+            "reference count overflow!"
+        );
+        refs == 1
+    }
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self {
+            spans: Arc::new(Slab::new()),
+        }
+    }
+}
+
+struct RegistryVisitor<'a>(&'a mut HashMap<&'static str, Value>);
+
+impl<'a> Visit for RegistryVisitor<'a> {
+    // TODO: special visitors for various formats that honeycomb.io supports
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        let s = format!("{:?}", value);
+        self.0.insert(field.name(), json!(s));
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: Data) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, Data>> {
+        self.spans.get(id_to_idx(id))
+    }
+
+    fn set_parent(&self, current: &Id, parent: &Id) {
+        // TODO(david): remove `.expect()`; return an error.
+        let current_span = self
+            .spans
+            .get(id_to_idx(current))
+            .expect(&format!("Missing current span with id: {:?}", current));
+        let parent_span = self
+            .spans
+            .get(id_to_idx(parent))
+            .expect(&format!("Missing parent span with id: {:?}", parent));
+
+        // current_span.parent = Some(parent.clone());
+        // parent_span.children.push(current.clone())
+    }
+
+    /// Decrements the reference count of the span with the given `id`, and
+    /// removes the span if it is zero.
+    ///
+    /// The allocated span slot will be reused when a new span is created.
+    fn drop_span(&self, id: Id) -> bool {
+        let idx = id_to_idx(&id);
+        if !self
+            .spans
+            .get(idx)
+            .and_then(|span| Some(span.drop_ref(Ordering::Relaxed)))
```

Comment:
This needs to be `Release`:
```suggestion
            .and_then(|span| Some(span.drop_ref(Ordering::Release)))
```
(but i really think it should be the `drop_ref` method that determines the ordering, so it's defined in one place...)

Reply:
fixed, thanks.